File: Array/Find all Missing Number/By set.java
// Method - Set(BruteForce)
// Time Complexity T - O(n)
// Space Complexity  S - O(1) 


class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        
        // initializing the missingNum variable
        List<Integer> missingNum = new ArrayList<>();
        
        // initializing the number varibale which makes copy of unique values from the nums
        Set<Integer> numbers = new HashSet<>();
        
        // looping through the array "0" to "n-1"
        for(int i = 0; i < nums.length; i++){
            // adding the unique number to the set
            numbers.add(nums[i]);
        }
        
        // looping through the array "1" to "n"
        for(int i = 1; i <= nums.length; i++){
        
            // check the presence of "i" in the set
            // if "i" is present in the set, ignore
            // if "i" is not present in the set, 
            // then insert the "i" in the missingNum
            if(!numbers.contains(i)){
                missingNum.add(i);    
            }
        }
        
        // return the mising numbers        
        return missingNum;
    }
}

File: Array/First Missing Positive Number/By set.cpp
// Method - Hashset(BruteForce)
// Time Complexity T - O(n)
// Space Complexity S - O(n)

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
       // creating a hash set to store the unique values
       set<int, greater<int>> sets;
        
        // loop runs for "0" to "n-1" times
        for(int i=0; i < nums.size(); i++)
        {
            
            // if "i" th value is 0, skip the iteration
            if(nums[i]<=0){
                continue;
            }
            
            // adding the values to the set
            sets.insert(nums[i]);
            
        }  
        
        // result variable to store the missing value
        int result = 1;
        
        // loop runs for 1 to "n" times
        for(int i = 1; i <= sets.size(); i++)
        {
            // ensure the "i"th element is present in the set
            // if not found, return the index
            if(sets.find(i) == sets.end()){
                return i;
            }
            // incrementing the result is "i" is found in the set
            result++;
        }
        return result;  
    }
};
